Add IPv6 for esp-idf framework

HeMan opened the pull request:
# What does this implement/fix? 

Add IPv6 support for esp-idf framework and enables `api` and `ota` to listen on IPv6 socket. Partly solves https://github.com/esphome/feature-requests/issues/718

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1749

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
network:
  enable_ipv6: true
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

agners commented on esphome/components/wifi/wifi_component_esp_idf.cpp:
```diff
@@ -502,6 +505,12 @@ std::string format_ip4_addr(const esp_ip4_addr_t &ip) {
+  char buf[50];
+
+  snprintf(buf, sizeof(buf), IPV6STR, IPV62STR(ip));
+  return buf;
```
You can't return `buf` here, buf is on the stack... Use the following to create a copy on the heap.

```suggestion
  return std::string(buf);
```

HeMan replied:
Should I modify the `format_ip4_addr` in the same way?

agners replied:
Uh yes, that seems to be a bug.

oxan replied:
Actually, it works fine because `buf` is a `char*` and the return type is `std::string`, so C++ does a copy initialization using the conversion constructor of `std::string`... but yes, being explicit about it is much better.

agners replied:
Oh I see. Today I learned :smile:

HeMan commented:
I don't think my PR enabling IPv6 should contain that change.

oxan commented on esphome/components/wifi/__init__.py:
```diff
@@ -253,6 +254,7 @@ def _validate(config):
             cv.Optional(CONF_MANUAL_IP): STA_MANUAL_IP_SCHEMA,
             cv.Optional(CONF_EAP): EAP_AUTH_SCHEMA,
             cv.Optional(CONF_AP): WIFI_NETWORK_AP,
+            cv.Optional(CONF_ENABLE_IPV6): cv.boolean,
```
This seems to only be implemented for ESP-IDF, so it should only be configurable when using IDF (might need reformatting):
```suggestion
            cv.SplitDefault(CONF_ENABLE_IPV6, esp32_idf=False): cv.All(cv.only_with_esp_idf, cv.boolean),
```

oxan commented on esphome/components/wifi/__init__.py:
```diff
@@ -349,6 +351,13 @@ async def to_code(config):
+    if CONF_ENABLE_IPV6 in config and config[CONF_ENABLE_IPV6]:
+        cg.add_build_flag(
+            "-DIPV6_ENABLE -DCONFIG_LWIP_IPV6 -DCONFIG_LWIP_IPV6_AUTOCONFIG"
+        )
```
Use three separate calls here, but I think the CONFIG_LWIP_* flags are better set through sdkconfig instead.

oxan commented on esphome/components/wifi/__init__.py:
```diff
@@ -349,6 +351,13 @@ async def to_code(config):
+            "-DIPV6_ENABLE -DCONFIG_LWIP_IPV6 -DCONFIG_LWIP_IPV6_AUTOCONFIG"
+        )
+        if CORE.is_esp8266:
+            cg.add_build_flag("-DPIO_FRAMEWORK_ARDUINO_LWIP2_IPV6_LOW_MEMORY")
```
This is unreachable then.

oxan reviewed (changes requested):
Can we implement this without a preprocessor define? We generally only use those to limit code size, but that doesn't seem relevant here.

HeMan commented:
> Can we implement this without a preprocessor define? We generally only use those to limit code size, but that doesn't seem relevant here.

It works if I change to always use `AF_INET6` and `sockaddr_in6` in setup, but debug logs show's IPv4 address in an IPv6-format:
```
[12:49:07][D][api:097]: Accepted ::FFFF:C0A8:2B3
[12:49:10][VV][api.socket:796]: ::FFFF:C0A8:2B3: Connection closed
[12:49:10][W][api.connection:082]: ::FFFF:C0A8:2B3: Connection closed
[12:49:10][V][api:109]: Removing connection to ::FFFF:C0A8:2B3
```

oxan commented:
Yeah, I think that's fine. We can always add a helper function to format it in traditional IPv4 format if it turns out to be a problem.

HeMan commented:
> > Can we implement this without a preprocessor define? We generally only use those to limit code size, but that doesn't seem relevant here.
> 
> It works if I change to always use `AF_INET6` and `sockaddr_in6` in setup, but debug logs show's IPv4 address in an IPv6-format:

Bugger, it didn't work with ESP8266 without IPv6 support.

oxan commented:
I've been thinking a bit more about this, and I think we should move the creation of a "wildcard" socket that listens on all addresses to the `socket` component. API is not the only place that creates a socket (see e.g. `ota`), and I don't want to have to put `#ifdef USE_IPV6` conditionals everywhere. Something like this:

```cpp
// functions in socket
std::unique_ptr<Socket> socket_ip(int type, int protocol);
void set_sockaddr_any(struct sockaddr *addr, uint16_t port);

// usage in api
socket_ = socket::socket_ip(SOCK_STREAM, 0);
struct sockaddr_storage addr;
socket::set_sockaddr_any((struct sockaddr *) &addr, this->port_);
socket_->bind((struct sockaddr *) &addr, sizeof(addr));
```
That function can then also properly set `sin6_addr` to `in6addr_any` instead of leaving it empty like it's now.

Furthermore, I think putting the `enable_ipv6` setting under wifi is the wrong choice. Enabling IPv6 is a global property of the TCP/IP stack, not only of the wifi stack. E.g. on ESP32 using ethernet we could also enable IPv6, but the whole wifi component isn't even loaded there. Instead, I think we should put it under the `network` component.

HeMan commented:
That looks like a good solution, I think it could also solve what I'm trying to accomplish in https://github.com/esphome/esphome/pull/2977 as well.

I can't find the `network` component, should I put it under core `esphome`?

oxan commented:
> I can't find the `network` component, should I put it under core `esphome`?

There is, see `esphome/components/network`. It doesn't currently have a config schema, but you can add it.

> Agreed, if not necessary (and currently there seems to be no reason), I'd try to use dual-stack socket if IPv6 is (compile-time) enabled and use regular IPv4 if not in our socket component, and abstract all away from higher level.

Yes, agreed, especially on abstracting it away.

> Afaik, `in6addr_any` is zero anyway, so if the struct was empty, it should be the same. But using it explicitly is nicer of course.

Ah yeah, you're right about that, so disregard that comment.

> That said, to avoid confusion, I think we should make sure to format those in IPv4 format when logging, otherwise we will end up with issue reports and whatnot :) LwIP has a macro to detect those: IP_IS_V4MAPPEDV6.

Yeah, agreed. Can be implemented in `std::string Socket::getpeername()`, I think.

agners commented:
> > Can we implement this without a preprocessor define? We generally only use those to limit code size, but that doesn't seem relevant here.
> 
> It works if I change to always use `AF_INET6` and `sockaddr_in6` in setup, but debug logs show's IPv4 address in an IPv6-format:

What works? Compiling without IPv6 and use `AF_INET6`? Looking at lwIP, this shouldn't be working

https://github.com/espressif/esp-lwip/blob/2.1.2-esp/src/include/lwip/sockets.h#L81

If it really works, I am actually worried what definition of `AF_INET6` and `sockaddr_in6` we are using :cold_sweat: 

> Can we implement this without a preprocessor define? We generally only use those to limit code size, but that doesn't seem relevant here.

In the last version @HeMan took this a bit too literary I think. @oxan given above findings, I don't think it will be possible entirely without preprocessor defines, I think? I'd say we should limit ifdef's to a generic socket abstraction layer (as you outlined in https://github.com/esphome/esphome/pull/2953#issuecomment-1005591195).

HeMan commented:
Just noted that esp-idf is always built with `CONFIG_LWIP_IPV6`, even without my changes. Adding `add_idf_sdkconfig_option("CONFIG_LWIP_IPV6", False)` to mainline will make it fail.

agners commented:
> Just noted that esp-idf is always built with `CONFIG_LWIP_IPV6`, even without my changes. Adding `add_idf_sdkconfig_option("CONFIG_LWIP_IPV6", False)` to mainline will make it fail.

Ok, that explains why it works using `AF_INET6`.

I guess the question is if that's the case for all configuration (ESP32 Arduino, ESP8266). If not, we probably still need to have the option.

oxan commented:
To be explicit about it, my goals on this topic are three-fold (aside from getting IPv6 working, obviously):
* Keep the behaviour of the BSD socket API as close to the standard as possible. It might not be the best designed API, and deviating from the standard can be convenient, but it's a well-known API with well-known behaviour, and changing that'll only hurt us in the future when it causes unexpected breakage because people don't expect us to behave different from the rest of the world.
* Move as much logic for compatibility with both IPv4 and IPv6 into the socket component, so that all consumers can be kept relatively simple. Where this can't be done using the BSD sockets API, we can add additional wrapper/helper functions.
* If it's possible, I would like to avoid introducing our own preprocessor define for IPv6 support. Using pre-existing defines from IDF/lwIP to check what's available is fine, as long as we can limit it to the socket component.

From that perspective, this PR is fine with me, except that it obviously depends on #3015 to keep working on IPv4, and see my comment there.

HeMan committed:
Added IPv6 option

HeMan committed:
Enable IPv6 address activation.

oxan commented on esphome/components/api/api_server.cpp:
```diff
@@ -43,11 +47,18 @@ void APIServer::setup() {
   server.sin_family = AF_INET;
   server.sin_addr.s_addr = ESPHOME_INADDR_ANY;
   server.sin_port = htons(this->port_);
+#endif
```
Can we move this (and the creation of the socket) into helper functions? We'll need to do this logic for any component support IPv6, and I don't want to have to litter `#ifdef IPV6_ENABLE` over the whole code.

oxan commented on esphome/components/network/__init__.py:
```diff
@@ -1,7 +1,35 @@
+CONFIG_SCHEMA = cv.All(
+    cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(NetworkComponent),
```
This isn't necessary (and incorrect, as there's no class `NetworkComponent` on the C++ side).

oxan commented on esphome/components/network/__init__.py:
```diff
@@ -1,7 +1,35 @@
+
+    if CONF_ENABLE_IPV6 in config and config[CONF_ENABLE_IPV6]:
+        add_idf_sdkconfig_option("CONFIG_LWIP_IPV6", True)
+        add_idf_sdkconfig_option("CONFIG_LWIP_IPV6_AUTOCONFIG", True)
```
We'll probably need to actually define `ENABLE_IPV6` as well... 

Please name it `USE_IPV6` and add it to `esphome/core/defines.h` as well.

HeMan committed:
Added helper functions for IPv6 sockets.

- This should simplify getting a socket and a server struct that could handle
  IPv6 without having to have compiler macros in every server component.

HeMan committed:
Return size of used servertype.

- Return size of server struct.

HeMan committed:
Converter api to use helpers.

- Changed the api component to use socket::socket_ip and
  socket::set_sockaddr_any. This makes it listen to IPv6 when enabled.

HeMan committed:
Converter ota to use helpers.

- Changed the ota component to use socket::socket_ip and
  socket::set_sockaddr_any. This makes it listen to IPv6 when enabled.

oxan committed:
Fixes/cleanup

oxan committed:
Remove unused improt

jesserockz reviewed (approved):
I don't have ipv6 in my network, but my test device still works if I enable this :tada:
